src/frontend/src/binder/insert.rs

```diff
@@ -135,151 +130,234 @@ impl Binder {
             }
         };
 
-        // When the column types of `source` query do not match `expected_types`, casting is
-        // needed.
-        //
-        // In PG, when the `source` is a `VALUES` without order / limit / offset, special treatment
-        // is given and it is NOT equivalent to assignment cast over potential implicit cast inside.
-        // For example, the following is valid:
-        // ```
-        //   create table t (v1 time);
-        //   insert into t values (timestamp '2020-01-01 01:02:03'), (time '03:04:05');
-        // ```
-        // But the followings are not:
-        // ```
-        //   values (timestamp '2020-01-01 01:02:03'), (time '03:04:05');
-        //   insert into t values (timestamp '2020-01-01 01:02:03'), (time '03:04:05') limit 1;
-        // ```
-        // Because `timestamp` can cast to `time` in assignment context, but no casting between them
-        // is allowed implicitly.
-        //
-        // In this case, assignment cast should be used directly in `VALUES`, suppressing its
-        // internal implicit cast.
-        // In other cases, the `source` query is handled on its own and assignment cast is done
-        // afterwards.
-        let (source, cast_exprs, nulls_inserted) = match source {
-            Query {
-                with: None,
-                body: SetExpr::Values(values),
-                order_by: order,
-                limit: None,
-                offset: None,
-                fetch: None,
-            } if order.is_empty() => {
-                let (values, nulls_inserted) =
-                    self.bind_values(values, Some(expected_types.clone()))?;
-                let body = BoundSetExpr::Values(values.into());
-                (
-                    BoundQuery {
-                        body,
-                        order: vec![],
-                        limit: None,
-                        offset: None,
-                        with_ties: false,
-                        extra_order_exprs: vec![],
-                    },
-                    vec![],
-                    nulls_inserted,
-                )
+        let (returning_list, fields) = self.bind_returning_list(returning_items)?;
+        let is_returning = !returning_list.is_empty();
+
+        if !target_col_idents.is_empty() {
```

Comment:
this line separate the queries into two cases
case 1:
```sql
insert into t (v2, v1, v3) values (1, 2, 'a'), (3, 4, 'b');
```
case 2:
```sql
insert into t values (1, 2, 'a'), (3, 4, 'b');
```

Reply:
I feel the original big let block is more compact, but maybe it's also the cause of the bug. 🤔 Is it possible to reuse some code in the new version?

Reply:
In other words, can we make it less spaghetti? 🙏🥹